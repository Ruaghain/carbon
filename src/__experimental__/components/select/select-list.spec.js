import React from 'react';
import { shallow, mount } from 'enzyme';
import SelectList from './select-list.component';

describe('SelectList', () => {
  const defaultOptions = [{
    value: '1',
    text: 'Orange',
    options: { additional: true }
  }, {
    value: '2',
    text: 'Blue',
    options: { additional: true }
  }, {
    value: '3',
    text: 'Red',
    options: { additional: true }
  }];

  const fakeTarget = {
    getBoundingClientRect: () => ({
      top: 100,
      height: 200,
      pageYOffset: 300,
      width: 400,
      left: 500
    })
  };

  const renderWrapper = ({
    props,
    options = defaultOptions,
    type = shallow
  } = {}) => type(
    <SelectList { ...props }>
      {
        options.map(item => <div { ...item } />)
      }
    </SelectList>
  );

  it('renders a ScrollableList inside of a Portal', () => {
    expect(renderWrapper()).toMatchSnapshot();
  });

  it('filters the items based on the filter value', () => {
    const wrapper = renderWrapper();
    expect(wrapper.find('ScrollableList').children().length).toEqual(3);
    wrapper.setProps({ filterValue: 'orange' });
    expect(wrapper.find('ScrollableList').children().length).toEqual(1);
  });

  it('renders a no results item if filter removed all children', () => {
    const props = { filterValue: 'x' };
    const wrapper = renderWrapper({ props });
    expect(wrapper.find('ScrollableList').children()).toMatchSnapshot();
  });

  it('renders a no results item if no childen are provided', () => {
    const wrapper = renderWrapper({ options: [] });
    expect(wrapper.find('ScrollableList').children()).toMatchSnapshot();
  });

  it('supports mouse events', () => {
    const events = {
      onMouseLeave: 'mouseLeave',
      onMouseEnter: 'mouseEnter',
      onMouseDown: 'mouseDown'
    };
    Object.keys(events).forEach((event) => {
      const props = { [event]: jest.fn() };
      const wrapper = renderWrapper({ props });
      wrapper.find('div').first().simulate(events[event]);
      expect(props[event]).toHaveBeenCalled();
    });
  });

  it('positions the list based on the target', () => {
    const props = { target: fakeTarget };
    const wrapper = renderWrapper({ props, type: mount });
    const style = 'left: 500px; top: 300px; width: 400px; position: absolute;';
    expect(wrapper.find('div').first().instance().getAttribute('style')).toEqual(style);
  });
});
