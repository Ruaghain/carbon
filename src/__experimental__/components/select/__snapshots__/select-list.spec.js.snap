// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectList renders a ScrollableList inside of a Portal 1`] = `
<Portal
  onReposition={[Function]}
>
  <div
    role="presentation"
  >
    <ScrollableList
      keyNavigation={true}
    >
      <styled.li
        id={
          Object {
            "additional": true,
            "text": "Orange",
            "value": "1",
          }
        }
        isSelectable={true}
        key=".0"
      >
        <div
          options={
            Object {
              "additional": true,
            }
          }
          text="Orange"
          value="1"
        />
      </styled.li>
      <styled.li
        id={
          Object {
            "additional": true,
            "text": "Blue",
            "value": "2",
          }
        }
        isSelectable={true}
        key=".1"
      >
        <div
          options={
            Object {
              "additional": true,
            }
          }
          text="Blue"
          value="2"
        />
      </styled.li>
      <styled.li
        id={
          Object {
            "additional": true,
            "text": "Red",
            "value": "3",
          }
        }
        isSelectable={true}
        key=".2"
      >
        <div
          options={
            Object {
              "additional": true,
            }
          }
          text="Red"
          value="3"
        />
      </styled.li>
    </ScrollableList>
  </div>
</Portal>
`;

exports[`SelectList renders a no results item if filter removed all children 1`] = `
<styled.li
  isSelectable={false}
>
  No results for "x"
</styled.li>
`;

exports[`SelectList renders a no results item if no childen are provided 1`] = `
<styled.li
  isSelectable={false}
>
  No results
</styled.li>
`;
